File: packages/javascript/tests/public-api.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { promises as fs } from 'fs';
import { join } from 'path';
import { Cachetta, hash } from 'cachetta';

/**
 * Black-box tests for the public `cachetta` API surface.
 *
 * These tests treat the package as a black box: they import only from
 * `cachetta` (no internal modules, no private attributes), exercise
 * the public API, and assert against observable behavior such as files
 * written to disk.
 */
describe('public hash() export', () => {
  let tempDir: string;

  beforeEach(async () => {
    tempDir = await fs.mkdtemp('cachetta-public-api-');
  });

  afterEach(async () => {
    await fs.rm(tempDir, { recursive: true, force: true });
  });

  // After issue #45 the decorator no longer rewrites a string `path` into a
  // `{name}-{hash}{ext}` sibling. The `hash` export remains the way to build
  // keyed file paths — consumers pass a `path:` callable that calls it.

  it('digest drives a user-built keyed file path', async () => {
    const cache = new Cachetta({
      path: (...args: unknown[]) => join(tempDir, `data-${hash(...args)}.json`),
    });
    const wrapped = cache(async (a: string, b: string) => ({ a, b }));

    await wrapped('x', 'y');

    const digest = hash('x', 'y');
    const expected = join(tempDir, `data-${digest}.json`);
    await fs.access(expected);
  });

  it('digest drives a user-built bare-hash file name (no extension)', async () => {
    const cache = new Cachetta({
      path: (...args: unknown[]) => join(tempDir, `cache-${hash(...args)}`),
    });
    const wrapped = cache(async (k: string) => k);

    await wrapped('k');

    const digest = hash('k');
    const expected = join(tempDir, `cache-${digest}`);
    await fs.access(expected);
  });
});

File: packages/python/tests/public_api_test.py
"""Black-box tests for the public ``cachetta`` API surface.

These tests treat the package as a black box: they import only from
``cachetta`` (no internal modules, no private attributes), exercise the
public API, and assert against observable behavior such as files
written to disk.
"""

import tempfile
from pathlib import Path

from cachetta import Cachetta, hash


def describe_hash_export():
    """The public ``hash`` export returns a stable digest consumers can
    use to build custom ``path=`` callables or external indexes — the
    same hasher cachetta used to embed into the auto-keyed sibling path
    before issue #45 removed that implicit behavior.
    """

    def test_digest_drives_user_built_keyed_file_path():
        with tempfile.TemporaryDirectory() as tmpdir:
            @Cachetta(path=lambda *a: f"{tmpdir}/data-{hash(*a)}.json")
            def fn(a, b):
                return {"a": a, "b": b}

            fn("x", "y")

            digest = hash("x", "y")
            expected = Path(tmpdir) / f"data-{digest}.json"
            assert expected.exists(), (
                "Expected the cache file to be written at %s; tmpdir contains %s"
                % (expected, list(Path(tmpdir).iterdir()))
            )

    def test_digest_drives_user_built_keyed_path_with_kwargs():
        with tempfile.TemporaryDirectory() as tmpdir:
            @Cachetta(path=lambda *a, **kw: f"{tmpdir}/data-{hash(*a, **kw)}.json")
            def fn(a, *, flag):
                return (a, flag)

            fn("a", flag=True)

            digest = hash("a", flag=True)
            assert (Path(tmpdir) / f"data-{digest}.json").exists()

    def test_digest_drives_user_built_subdirectory_layout():
        with tempfile.TemporaryDirectory() as tmpdir:
            cache_dir = Path(tmpdir) / "cache"

            @Cachetta(path=lambda *a: cache_dir / hash(*a))
            def fn(k):
                return k

            fn("k")

            digest = hash("k")
            assert (cache_dir / digest).exists()
